### ISSUE TITLE
Best practise in building a large GraphQL server with minimal code duplication

### ISSUE THREAD

**Issue by jakelacey2012** [apollostack/apollo#62]
So what do I mean by a large GraphQL server? - 
 - It has many many connectors 5+ (Mongo, Redis, MySQL, ProSQL, ....etc... and counting)
 - Many models and schemas, (which has duplicate definitions)
 - and many duplicate resolvers.

So I just want to get opinions, on how someone would structure their code to reduce duplication.

Thanks!!

**Comment by jbaxleyiii** [comment]
@jakelacey2012 I'm not sure if this is the best way to do it ;) But this is our production server: https://github.com/NewSpring/Heighliner

We have been pretty happy with it structure wise

**Comment by helfer** [comment]
Also, in terms of structure, here's our current view: https://github.com/apollostack/graphql-tools/blob/master/connectors.md

There's also a Medium post about a similar topic: https://medium.com/apollo-stack/how-to-build-graphql-servers-87587591ded5

**Comment by jakelacey2012** [comment]
Sweet, I'll have a read - thanks guys!!

**Comment by jakelacey2012** [comment]
Thanks again guys for getting back to me, I really appreciate it!!

Heighliner looks very interested, we made a prototype of a very similar solution creating the schemas, resolvers and models in separate files scoping them as different data sources and joining them together to define the entire graphQL server.

@helfer thanks for the posts, and that's how I've developed graphQL servers in the past, however a friend and I were discussing how to build a framework to make the developer experience better. (You must has been asked this 50x) is this framework something your aiming to build in Apollo servers?

Thanks :)

**Comment by helfer** [comment]
@jakelacey2012 Yeah, we know that building GraphQL servers isn't easy. We're not working on building such a framework at the moment, but I think we'll get there a little bit down the road, when we've solved what we consider bigger blockers to adoption at the moment (specifically GraphQL clients).

I think at the very least, Apollo Server should be a good foundation to build a set of tools on that would fulfill much of the same purpose as a framework. If you want to start building some of those things (like connectors, authentication helpers, decorators etc.), I'd be happy to help where I can!
